R/equals_na_lintr.R

```diff
@@ -1,38 +1,22 @@
-#' @describeIn linters that checks for x == NA
+#' @describeIn linters that checks for x == NA and x != NA
 #' @export
 equals_na_linter <- function(source_file) {
-  all_matches <- re_matches(
-    source_file$lines,
-    rex("==", zero_or_more(" "), "NA"),
-    locations = TRUE,
-    global = TRUE
-  )
-  line_numbers <- as.integer(names(source_file$lines))
 
-  Map(
-    function(line_matches, line_number) {
-      lapply(
-        split(line_matches, seq_len(nrow(line_matches))),
-        function(match) {
-          start <- match[["start"]]
-          if (is.na(start)) {
-            return()
-          }
-          end <- match[["end"]]
-          Lint(
-            filename = source_file$filename,
-            line_number = line_number,
-            column_number = start,
-            type = "warning",
-            message = "Use is.na rather than == NA.",
-            line = source_file$lines[[as.character(line_number)]],
-            ranges = list(c(start, end)),
-            linter = "equals_na_linter"
-          )
-        }
-      )
-    },
-    all_matches,
-    line_numbers
-  )
+  if (!length(source_file$parsed_content)) return(list())
+
+  xml <- source_file$xml_parsed_content
+
+  comparators <- c("EQ", "NE")
+  comparator_table <- paste0("self::", comparators, collapse = " or ")
+  NA_values <- c("NA", "NA_integer_", "NA_real_", "NA_complex_", "NA_character_")
+  NA_table <- paste("text() =", sQuote(NA_values, "'"), collapse = " or ")
+
+  xpath_fmt <- "//expr[expr[NUM_CONST[%s]]]/*[%s]"
+  xpath <- sprintf(xpath_fmt, NA_table, comparator_table)
+
+  bad_expr <- xml2::xml_find_all(xml, xpath)
+
+  lapply(bad_expr, xml_nodes_to_lint, source_file,
```

Comment:
The implementation `xml_nodes_to_lint` is missing from this PR.
Could you please add it?

Reply:
totally forgot to include that edit to utils.R earlier 😅